packages/hierarchy-builder/src/hierarchy-builder/TreeNodesProvider.ts

```diff
@@ -0,0 +1,515 @@
+    this._directNodesCache = new Map();
+  }
+
+  private loadDirectNodes(parentNode: TreeNode | undefined): Observable<InProgressTreeNode> {
+    const enableLogging = false;
+    return from(this._queryBuilder.createQueries(parentNode)).pipe(
+      mergeMap((queries) => from(queries)),
+      (queryObservable) =>
+        this._scheduler.scheduleSubscription(
+          queryObservable.pipe(
+            tap(() => enableLogging && console.log(`[loadDirectNodes] Do query for ${parentNode ? JSON.stringify(parentNode) : "<root>"}`)),
+            mergeMap((query) => from(this._queryReader.read(this._queryExecutor, { ...query, ecsql: applyLimit(query.ecsql, query.ctes) }))),
+          ),
+        ),
+      mergeAll(),
+      map(convertECInstanceIdSuffixToBase36),
+      shareReplay({ refCount: false }),
+    );
+  }
+
+  private ensureDirectChildren(parentNode: TreeNode | undefined): Observable<InProgressTreeNode> {
+    const enableLogging = false;
+    const key = parentNode ? JSON.stringify(parentNode.key) : "";
+
+    const cached = this._directNodesCache.get(key);
+    if (cached) {
+      enableLogging && console.log(`[ensureDirectChildren] Found direct nodes observable for ${parentNode ? parentNode.label : "<root>"}`);
+      return cached;
+    }
+
+    const obs = this.loadDirectNodes(parentNode);
+    this._directNodesCache.set(key, obs);
+    enableLogging && console.log(`[ensureDirectChildren] Saved direct nodes observable for ${parentNode ? parentNode.label : "<root>"}`);
+    return obs;
+  }
+
+  public getNodesObservable(parentNode: TreeNode | undefined): Observable<InProgressTreeNode> {
+    if (parentNode && Array.isArray(parentNode.children)) {
+      return from(parentNode.children);
+    }
+
+    const directChildren = this.ensureDirectChildren(parentNode);
+    const result = directChildren.pipe(
+      createMergeInstanceNodesByLabelReducer(this._directNodesCache),
+      createHideIfNoChildrenReducer((n) => this.hasNodes(n), false),
+      createHideNodesInHierarchyReducer(this, this._directNodesCache, false),
+      sortNodesByLabelReducer,
+      createClassGroupingReducer(this._schemas),
+    );
+    return parentNode ? result.pipe(createPersistChildrenReducer(parentNode)) : result;
+  }
+
+  public async getNodes(parentNode: TreeNode | undefined): Promise<TreeNode[]> {
+    return new Promise((resolve, reject) => {
+      const nodes = new Array<TreeNode>();
+      this.getNodesObservable(parentNode)
+        // finalize before returning
+        .pipe(createDetermineChildrenReducer(this), supplyIconsReducer)
+        // load all nodes into the array and resolve
+        .subscribe({
+          next(node) {
+            nodes.push(node);
+          },
+          error(err) {
+            reject(err);
+          },
+          complete() {
+            resolve(nodes);
+          },
+        });
+    });
+  }
+
+  public hasNodes(node: TreeNode): Observable<boolean> {
+    const enableLogging = false;
```

Comment:
I see `enableLogging` in majority of functions in this file. Does it usually worth logging just in subset of them? If not then maybe there could be single variable at file level?

Reply:
Having it at file level makes it very hard to debug one specific reducer due to all the noise coming from other reducers.